File: problems/two-sum/solution1.py
class BruteForce:
    @staticmethod
    def two_sum(nums: list[int], target: int) -> list[int] | None:
        for i, num in enumerate(nums):
            complement = target - num
            if complement in nums[i + 1:]:
                return [i, nums.index(complement, i + 1)]
        return None


class SinglePass:
    @staticmethod
    def two_sum(nums: list[int], target: int) -> list[int] | None:
        # remaining to index mapping
        seen = {}

        for i, num in enumerate(nums):
            complement = target - num
            if complement in seen:
                return [seen[complement], i]
            seen[num] = i
        return [-1, -1]

File: problems/two-sum/Solution1.java
import java.util.HashMap;

public class Solution1 {
  public static class BruteForce {
    public static int[] twoSum(int[] nums, int target) {
      for (var i = 0; i < nums.length; i++) {
        for (var j = i + 1; j < nums.length; j++) {
          if (nums[j] == target - nums[i]) {
            return new int[] {i, j};
          }
        }
      }
      return null;
    }
  }

  public static class SinglePass {
    public static int[] twoSum(int[] nums, int target) {
      // remaining to index mapping
      final var indices = new HashMap<Integer, Integer>();

      for (int i = 0; i < nums.length; i++) {
        final int num = nums[i];
        if (indices.containsKey(num)) {
          return new int[]{indices.get(num), i};
        }
        indices.put(target - num, i);
      }
      return new int[]{-1, -1};
    }
  }
}
